File: exemplos_aula09/java/ex2_sem.java
import java.util.concurrent.*;

public class ex2_sem {
	static int num_threads = 10;
	static volatile int r = 0;

	static class ex2_thread implements Runnable {
		Semaphore s;
		private int tid;

		public ex2_thread(int id, Semaphore sem){
			this.tid = id;
			s = sem;
		}

		public void run(){
			while (true) {
				try {
					s.acquire();
					r++;
					System.out.println("thread " + tid + " acessando recurso.. (" + r + ")");
					Thread.sleep(100);
					System.out.println("thread " + tid + " liberando recurso.. (" + r + ")");
					r--;
					s.release();
					Thread.sleep(100);
				}catch (InterruptedException e) {
				}
			}
		}
	}

	public static void main(String[] args) {
		Semaphore sem = new Semaphore(3);
		int i;

		for (i = 0; i < num_threads; i++){
			Thread t = new Thread(new ex2_thread(i, sem));
			t.start();
		}

	}
}

File: exemplos_aula09/c/ex2_sem.c
#include <stdio.h>
/*
exemplo com semaforos

no maximo 3 threads estao acessando o recurso ao mesmo tempo.
*/

#include <pthread.h>
#include <semaphore.h>

#define THREADS		10

sem_t sem;
int r = 0;

void *task(void *arg){
	int item, tid, i;

	tid = (int)(long int)arg;
	while(1){
		sem_wait(&sem);
		r++;
		printf("thread %d acessando recurso.. (%d)\n", tid, r);
		usleep(100000);
		printf("thread %d liberando recurso.. (%d)\n", tid, r);
		r--;
		sem_post(&sem);
		usleep(100000 * THREADS);
	}
}

int main(void){
	long int i;
	pthread_t threads[THREADS];
	
	sem_init(&sem,0,3);

	for(i = 0; i < THREADS; i++)
		pthread_create(&threads[i], NULL, task, (void *)i);

	pthread_exit(NULL);
	return(0);
}

